File: leetcode_mixed/largestHistogram.cpp
/* 
Problem Statement : https://leetcode.com/problems/largest-rectangle-in-histogram/

Approach1 :
step 1) for each bar cal the next smaller bar index at right and left indeex
step 2) Calculate area = bar_height * (right - left - 1) 
Time Complexity: O(N^2)
Space Complexity: O(1) not using auxilarry space

Approach2 : 
step 1) Calcluate next smaller for all.
step 2) Calcluate next previous smaller for all.
step 3) now, Calculate area of rectangle for each bar, and pick the largest from them.

Time Complexity:  O(N) in three passes
Space Complexity: O(N)
*/
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

class Solution {
private:
    void getNextSmaller(vector <int> heights, vector <int> &arr, int size){
        
        stack <int> st;
        st.push(-1);
        
        //get the indices for all next smaller for each heihts bar
        
        for (int i = size - 1 ; i >= 0 ; i--){
            
            //stack(top) == -1 means no such next element is smaller for this current
            while(st.top() !=-1 and heights[st.top()] >= heights[i]) 
                st.pop();
            
            ////We are pushing indices instead of value  since we are more interested in indices that will help to calculate the ractangle width.
            arr[i] = st.top();
            st.push(i);    
        }
    }
    
    void getPrevSmaller(vector <int> heights, vector <int> &arr, int size){
        
        stack <int> st;
        st.push(-1);
        
        //get the indices for all next smaller for each heihts bar
        
        for (int i = 0 ; i < size ; i++){
            
            //stack(top) == -1 means no such prev element is smaller for this current 
            while(st.top() !=-1 and heights[st.top()] >= heights[i]) 
                st.pop();
            
            //We are pushing indices instead of value 
            arr[i] = st.top();
            st.push(i);     
        }
    }
    
    
public:
    int largestRectangleArea(vector<int>& heights){
        
        int size = heights.size();
        vector <int> right(size);
        vector <int> left(size);
        
        //get the indices for all next smaller
        getNextSmaller(heights, right, size);
        
        //get the indices for all prev smaller
        getPrevSmaller(heights, left, size);
        
        int maxArea = 0;
        
        for(int i = 0; i < size; i++){
            
            //calculate each bar Area
            int area = heights[i] * ((right[i] == -1 ? size : right[i]) - left[i] -1); //if right index '-1' i.e. right index we have to make 'size' that means for current bar no such element of its right is lesser.  
            maxArea = max(area, maxArea);
            
        }
        return maxArea;
    }
};

File: stack_Queue/largestHistogram.py
"""
Problem Statement : https://leetcode.com/problems/largest-rectangle-in-histogram/submissions/
Time Complexity: O(N)
Space Complexity: O(N)

Approach using the two stack.
"""

class Solution(object):
    def nearestLeft(self, arr, size):
        #this steps help us to pop element from stack after for loop
        nleft = [0] * size
        stack = []

        for prev in range(size-1, -1, -1):
            #when stack is not empty and prev element is small than stack.top
            
            while stack and arr[prev] < arr[stack[-1]]:
                nleft[stack[-1]] = prev+1
                stack.pop()

            stack.append(prev)
        
        #it'll return the nerset max left
        return nleft
    
    
    def nearestRight(self, arr, size):
        #this steps help us to pop element from stack after for loop
        nright = [size-1] * size
        stack = []
        
        for next in range(size):

            #when stack is not empty and next element is small than stack.top
            while stack and arr[next] < arr[stack[-1]]:
                nright[stack[-1]] = next - 1
                stack.pop()

            stack.append(next)

        #it'll return the nerset max right
        return nright
    
    def largestRectangleArea(self, heights):
        size = len(heights)
        #get the nearest maximum at left
        left = self.nearestLeft(heights, size)
        #print(left)
        
        #get the nearest maximum at right
        right = self.nearestRight(heights, size)
        
        #print(right)
        max_area = 0
        for i in range(size):
            area = heights[i] * (right[i] - left[i] + 1)
            max_area = max(max_area, area)

        return max_area